File: src/auth/auth.service.ts
import { Injectable, ConflictException, UnauthorizedException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { JwtService } from '@nestjs/jwt';
import * as bcrypt from 'bcryptjs';
import { User, UserRole } from './user.entity';
import { RegisterDto, LoginDto } from './dto/auth.dto';

@Injectable()
export class AuthService {
  constructor(
    @InjectRepository(User)
    private readonly userRepo: Repository<User>,
    private readonly jwtService: JwtService,
  ) {}

  async onModuleInit() {
    const count = await this.userRepo.count();
    if (count === 0) {
      const hashed = await bcrypt.hash('admin123', 10);
      await this.userRepo.save(this.userRepo.create({
        name: 'Admin',
        email: 'admin@campus.com',
        password: hashed,
        role: UserRole.ADMIN,
      }));
      console.log('Seeded admin account: admin@campus.com / admin123');
    }
  }

  async register(dto: RegisterDto) {
    const existing = await this.userRepo.findOne({ where: { email: dto.email } });
    if (existing) {
      throw new ConflictException('Email already registered');
    }

    const hashed = await bcrypt.hash(dto.password, 10);
    const user = this.userRepo.create({
      name: dto.name,
      email: dto.email,
      password: hashed,
      role: UserRole.STUDENT,
    });

    const saved = await this.userRepo.save(user);
    const token = this.generateToken(saved);

    return {
      user: { id: saved.id, name: saved.name, email: saved.email, role: saved.role },
      token,
    };
  }

  async login(dto: LoginDto) {
    const user = await this.userRepo.findOne({ where: { email: dto.email } });
    if (!user) {
      throw new UnauthorizedException('Invalid credentials');
    }

    const valid = await bcrypt.compare(dto.password, user.password);
    if (!valid) {
      throw new UnauthorizedException('Invalid credentials');
    }

    const token = this.generateToken(user);

    return {
      user: { id: user.id, name: user.name, email: user.email, role: user.role },
      token,
    };
  }

  private generateToken(user: User): string {
    return this.jwtService.sign({ sub: user.id, email: user.email, role: user.role });
  }
}

File: src/main/java/com/courseenrollment/auth/service/AuthService.java
package com.courseenrollment.auth.service;

import com.courseenrollment.auth.dto.AuthResponse;
import com.courseenrollment.auth.dto.LoginRequest;
import com.courseenrollment.auth.dto.RegisterRequest;
import com.courseenrollment.auth.dto.UserDto;
import com.courseenrollment.auth.entity.User;
import com.courseenrollment.auth.enums.UserRole;
import com.courseenrollment.auth.repository.UserRepository;
import com.courseenrollment.common.exception.ConflictException;
import com.courseenrollment.config.JwtService;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;

    public AuthService(UserRepository userRepository,
                       PasswordEncoder passwordEncoder,
                       JwtService jwtService) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
        this.jwtService = jwtService;
    }

    @Transactional
    public AuthResponse register(RegisterRequest req) {
        if (userRepository.existsByEmail(req.getEmail())) {
            throw new ConflictException("Email already registered");
        }

        UserRole role;
        if (req.getRole() != null && !req.getRole().trim().isEmpty()) {
            role = UserRole.fromValue(req.getRole().trim());
        } else {
            // Default to ADMIN so all users can manage courses and students immediately
            role = UserRole.ADMIN;
        }

        String hashedPassword = passwordEncoder.encode(req.getPassword());
        User user = new User(req.getEmail().trim(), req.getName().trim(), hashedPassword, role);
        User savedUser = userRepository.save(user);

        String token = jwtService.generateToken(
                savedUser.getId(),
                savedUser.getEmail(),
                savedUser.getRole().getValue()
        );

        return new AuthResponse(UserDto.fromEntity(savedUser), token);
    }

    @Transactional
    public AuthResponse switchRole(String email, String targetRole) {
        User user = userRepository.findByEmail(email.trim())
                .orElseThrow(() -> new BadCredentialsException("User not found"));

        UserRole newRole = (targetRole != null && !targetRole.trim().isEmpty())
                ? UserRole.fromValue(targetRole.trim())
                : (user.getRole() == UserRole.ADMIN ? UserRole.STUDENT : UserRole.ADMIN);

        user.setRole(newRole);
        User savedUser = userRepository.save(user);

        String token = jwtService.generateToken(
                savedUser.getId(),
                savedUser.getEmail(),
                savedUser.getRole().getValue()
        );

        return new AuthResponse(UserDto.fromEntity(savedUser), token);
    }

    @Transactional(readOnly = true)
    public AuthResponse login(LoginRequest req) {
        User user = userRepository.findByEmail(req.getEmail().trim())
                .orElseThrow(() -> new BadCredentialsException("Invalid credentials"));

        if (!passwordEncoder.matches(req.getPassword(), user.getPassword())) {
            throw new BadCredentialsException("Invalid credentials");
        }

        String token = jwtService.generateToken(
                user.getId(),
                user.getEmail(),
                user.getRole().getValue()
        );

        return new AuthResponse(UserDto.fromEntity(user), token);
    }
}
